src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                // This also preserves the behavior the VB had through Dev12.
+                hardSelect = bestFilterResult.Value.VSCompletionItem.FilterText.StartsWith(filterText, StringComparison.CurrentCultureIgnoreCase);
+                index = filterResults.IndexOf(bestFilterResult.Value);
+            }
+            else
+            {
+                index = 0;
+                hardSelect = false;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, index, filters,
+                hardSelect ? AsyncCompletionData.UpdateSelectionHint.Selected : AsyncCompletionData.UpdateSelectionHint.SoftSelected,
+                centerSelection: true, uniqueItem: null);
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleAllItemsFilteredOut(
+            AsyncCompletionData.CompletionTriggerReason triggerReason,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            bool hasActiveFilters,
+            CompletionRules completionRules)
+        {
+            AsyncCompletionData.UpdateSelectionHint selection;
+            if (triggerReason == AsyncCompletionData.CompletionTriggerReason.Insertion)
+            {
+                // If the user was just typing, and the list went to empty *and* this is a 
+                // language that wants to dismiss on empty, then just return a null model
+                // to stop the completion session.
+                if (completionRules.DismissIfEmpty)
+                {
+                    return null;
+                }
+
+                selection = AsyncCompletionData.UpdateSelectionHint.NoChange;
+            }
+            else
+            {
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                selection = hasActiveFilters ? AsyncCompletionData.UpdateSelectionHint.NoChange : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                filters, selection, centerSelection: true, uniqueItem: default);
+        }
+
+        private IEnumerable<AsyncCompletionData.CompletionItemWithHighlight> GetHighlightedList(
+            IEnumerable<ExtendedFilterResult> filterResults,
+            string filterText,
+            bool highlightMatchingPortions)
```

Comment:
is this being called when generating the entire list of completion items?  if so, it will likely be too slow to do that.  that's how we started out, and we had to change things so that computing this was deferred until the completion list was showed and *only* computed for the items in display (even then we regressed some performance, but got an exception).